utils/preferences/tests/dtf/mod.rs

```diff
@@ -0,0 +1,84 @@
+mod data_provider;
+mod options;
+mod preferences;
+
+use data_provider::get_defaults;
+use icu_datetime::options::length;
+use icu_datetime::options::preferences::HourCycle;
+use icu_locid::{extensions_unicode_key, LanguageIdentifier, Locale};
+use icu_preferences::{preferences, Preferences};
+use options::{DayPeriod, LocaleMatcher};
+use preferences::{Calendar, NumberingSystem};
+
+preferences!(
+    DTFPreferencesBag,
+    DTFDefaultPreferencesBag,
```

Comment:
Nit: I'd prefer if we could get this down to 2 types. I think you don't need the `DTFDefaultPreferencesBag` type; I'd get rid of it.

Reply:
I think it's useful to store the default data for each locale. I could use `DTFResolvedPreferences` but that would require me to store a locale as value and as key. I'd like to consider each crate to have `DefaultPreferences` for DP to provide, and ResolvedPreferences to be a mix of default and optional preferences.

Reply:
I guess the way I see it, there are really only two types you need:

```rust
struct InputPreferences {
    hour_cycle: Option<HourCycle>,
    numbering_system: Option<NumberingSystem>,
    // ...
}

struct ResolvedPreferences {
    hour_cycle: HourCycle,
    numbering_system: NumberingSystem,
    // ...
}
```

I would love to have just 1 struct, but I'll accept that we need 2 since one has more granularity of options.

The only other difference between your structs is that they contain a LanguageIdentifier field. Can we avoid the extra types by using a tuple of LanguageIdentifier with the correct version of the preferences struct?

Reply:
I think that depends on how it'll be implemented in DataProvider.

I see three types. What you call `Input` preferences, `Resolved` preferences and `Default` preferences. You could fold `Default` into `Resolved` but it imposes two architectural suboptimal choices:

1) It makes DP data require to contain language identifier in a value, while it is not needed in DP, it's just needed in the struct.
2) It means that every resolved preference needs to have an encoded default value. This is not a requirement in architectural sense - we could have a preference that has a mandatory resolved value that is *calculated* in the constructor, rather than retrieved from some storage of default preference values.

Generally, I have a slight preference to separate default from resolved, and I am not convinced by an argument of reduction of types - they're in a macro and easy to maintain in result.

Reply:
Don't know if this is a good idea or not, but: What do you think about having two non-langid inner preference structs generated by the macro, one with `Option` on the fields and one without, and then a single `InputPreferences<T>`, `ResolvedPreferences<T>`, and `DefaultPreferences<T>` that adds in the langid in the appropriate form

Reply:
It seems reasonable! I'd need to think of what are implications on memory layout and ergonomics. @Manishearth do you have any API impact hunch here?

Reply:
I find multiple vaguely-similar types to be annoying but we already do that kind of thing a bunch and I think it's fine if documented well. I think this is fine.